governator-core/src/main/java/com/netflix/governator/internal/AbstractLifecycleFeature.java

```diff
@@ -1,55 +1,71 @@
 package com.netflix.governator.internal;
 
+import static com.netflix.governator.internal.AbstractLifecycleFeature.TypeVisitor.ElementType.*;
+
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
 
 import com.netflix.governator.LifecycleAction;
 import com.netflix.governator.LifecycleFeature;
 
 /**
- * Base utility class for field and method based lifecycle features.  This class
- * deals with identifying all the fields and methods for the type and its superclasses
- * and interfaces and delegating the actual processing using the template methods
- * getFieldAction and getMethodAction.
+ * Base utility class for field and method based lifecycle features. This class
+ * deals with identifying all the fields and methods for the type and its
+ * superclasses and interfaces and delegating the actual processing using the
+ * template methods getFieldAction and getMethodAction.
  * 
  * @author elandau
  */
 abstract class AbstractLifecycleFeature implements LifecycleFeature {
+    interface TypeVisitor {
```

Comment:
I like the idea of using a visitor here but this visitor doesn't exactly conform to the standard pattern.  

How about a basic visitor like this,

```java
public static class VisitableType {
    public interface Visitor {
        void visitMethod(Method method);
        void visitField(Field field);
        void visitSuper(Class<?> type);
        void visitInterface(Class<?> type);
    }
    
    private final Class<?> type;
    
    public VisitableType(Class<?> type) {
        this.type = type;
    }
    
    <T extends Visitor> T accept(T visitor) {
        if (type == null) 
            return visitor;
        
        for (final Field field : type.getDeclaredFields()) {
            visitor.visitField(field);
        }
 
        for (final Method method : type.getDeclaredMethods()) {
            visitor.visitMethod(method);
        }
 
        visitor.visitSuper(type.getSuperclass());

        for (Class<?> iface : type.getInterfaces()) {
            visitor.visitInterface(iface);
        }
        
        return visitor;
    }
}
```

The PostConstructVisitor would look something like this,

```java
public static class PostConstructVisitor implements VisitableType.Visitor {
    private List<LifecycleAction> postConstruct = new ArrayList<>();
    private final boolean includeSupersAndInterfaces;

    public PostConstructVisitor(boolean includeSupersAndInterfaces) {
         this.includeSupersAndInterfaces = includeSupersAndInterfaces;
    }

    @Override
    public void visitMethod(Method method) {
        // For brevity I'm omitting all the additional processing and checks
        if (method.getAnnotation(PostConstruct.class) != null) {
            postConstruct.add(new LifecycleAction() {
                @Override
                public void call(Object obj) throws Exception {
                    method.invoke(obj);
                }
            });
        }
    }

    @Override
    public void visitField(Field field) {
    }

    @Override
    public void visitSuper(Class<?> type) {
        if (includeSupersAndInterfaces && !hasPostConstruct()) {
            new VisitableType(type).accept(this);
        }
    }

    @Override
    public void visitInterface(Class<?> type) {
        if (includeSupersAndInterfaces && !hasPostConstruct()) {
            new VisitableType(type).accept(this);
        }
    }
    
    private boolean hasPostConstruct() {
        return !postConstruct.isEmpty();
    }
    
    public List<LifecycleAction> getActions() {
        return postConstruct;
    }
}
```

Usage is simply this,

```java
List<LifecycleActions> actions = new VisitableType(SomeType.class).accept(new PostConstructVisitor()).getActions();
```

You can then further customize PostConstructVisitor to adhere to either the existing behavior or that of the JSR.

Note that even though there may be additional overhead for iterating all elements of the class that's fine because the results are cached when the type is first encountered.